### ISSUE TITLE
discussion: create 0050 e10s

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#82]
we currently have (and it will need a tidy up/check for FF53, I think some of these may have fallen off be not longer applicable, eg 2652 probably won't apply).

EDIT: doh! of course 2652 will apply until e10s is enforced! - end edit

```js
/* 2650: start the browser in e10s mode (FF48+)
 * After restarting the browser, you can check whether it's enabled by visiting
 * about:support and checking that "Multiprocess Windows" = 1
 * use force-enable and extensions.e10sblocksenabling if you have add-ons ***/
   // user_pref("browser.tabs.remote.autostart", true);
   // user_pref("browser.tabs.remote.autostart.2", true); // (FF49+)
   // user_pref("browser.tabs.remote.force-enable", true); // (hidden pref)
   // user_pref("extensions.e10sBlocksEnabling", false);
/* 2651: control e10s number of container processes
 * [1] http://www.ghacks.net/2016/02/15/change-how-many-processes-multi-process-firefox-uses/
 * [2] https://bugzilla.mozilla.org/show_bug.cgi?id=1207306 ***/
   // user_pref("dom.ipc.processCount", 4);
/* 2652: enable console shim warnings for extensions that don't have the flag 'multiprocessCompatible' as true ***/
user_pref("dom.ipc.shims.enabledWarnings", true);
/* 2660: enforce separate content process for file://URLs (FF53+?)
 * [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1147911
 * [2] http://www.ghacks.net/2016/11/27/firefox-53-exclusive-content-process-for-local-files/ ***/
user_pref("browser.tabs.remote.separateFileUriProcess", true);
```

there is also this from FF53 stable
```js
/* enable WebExtension addon code to run in a separate process (webext-oop) (FF53+)
 * [1] https://wiki.mozilla.org/WebExtensions/Implementing_APIs_out-of-process */
   // pref("extensions.webextensions.remote", false);
```

and to add for info
```js
security.sandbox.content.level
```

This makes sense to put right at the start of the js, and it nicely takes a chunk out of the 2600's which is rather big and messy. Note: we still have to deal with containers later on as well in a new section.

**Comment by Thorin-Oakenpants** [comment]
quick and dirty early draft

```js
/*** 0050: MULTI-PROCESS (e10s) / SANDBOXING
     We recommend you let Firefox handle this. Until e10s is enforced, if
     - all your add-ons have the 'multiprocessCompatible' flag as true, then FF = e10s
     - any add-ons have 'multiprocessCompatible' flag as false, then FF != e10s
     - any add-ons missing the 'multiprocessCompatible' flag will be disabled (FF53=)
 ***/
/* 0051: start the browser in e10s mode (FF48+)
 * After restarting the browser, you can check whether it's enabled by visiting
 * about:support and checking that "Multiprocess Windows" = 1
 * use force-enable and extensions.e10sblocksenabling if you have add-ons ***/
   // user_pref("browser.tabs.remote.autostart", true);
   // user_pref("browser.tabs.remote.autostart.2", true); // (FF49+) (hidden pref)
   // user_pref("browser.tabs.remote.force-enable", true); // (hidden pref)
   // user_pref("extensions.e10sBlocksEnabling", false);
/* 0052: control e10s number of container processes
 * [1] http://www.ghacks.net/2016/02/15/change-how-many-processes-multi-process-firefox-uses/
 * [2] https://bugzilla.mozilla.org/show_bug.cgi?id=1207306 ***/
   // user_pref("dom.ipc.processCount", 4);
/* 0053: set sandbox level. DO NOT MEDDLE WITH THESE. They are included to inform you NOT to play
 * with them. The values are integers, but the code below deliberately contains a data mismatch
 * [1] https://wiki.mozilla.org/Sandbox
 * [2] http://www.ghacks.net/2017/01/23/how-to-change-firefoxs-sandbox-security-level/#comment-4105173 */
   // user_pref("security.sandbox.content.level", "donotuse")
   // user_pref("dom.ipc.plugins.sandbox-level.default", "donotuse");
   // user_pref("dom.ipc.plugins.sandbox-level.flash, "donotuse");
/* 0060: enable console shim warnings for add-ons with the 'multiprocessCompatible' flag as false ***/
user_pref("dom.ipc.shims.enabledWarnings", true);
/* 0061: enable WebExtension add-on code to run in a separate process (webext-oop) (FF53+)
 * [1] https://wiki.mozilla.org/WebExtensions/Implementing_APIs_out-of-process */
   // pref("extensions.webextensions.remote", true);
/* 0062: enforce separate content process for file://URLs (FF53+)
 * [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1147911
 * [2] http://www.ghacks.net/2016/11/27/firefox-53-exclusive-content-process-for-local-files/ ***/
   // user_pref("browser.tabs.remote.separateFileUriProcess", true);
```

**Comment by earthlng** [comment]
There's only one item for sandboxing and it's only to inform users. I'd prefer if you create a New Issue with FYI label and list the Prefs and URLs there, and we remove that item from the section and make the section only for e10s. 
Idk if it really needs to be such a low number like 0050, because once e10s is ready and default enabled everywhere + only WebExtensions are allowed, the only thing that someone maybe needs/wants to change is the process count number. I'd even say it could go under Personal, maybe as 3100 or so.
If you want to keep the 0050 then it shouldn't have 006x.
0061 - idk if this is ready to use just yet. Maybe better to keep this in our sticky to keep an eye on.
0061+2 both rely on e10s so I don't see why they have a new sub-number ie. 006x instead of 005x.
0061 - copypasta: pref instead of user_pref

The rest is great! very nice section description and all the item titles and links and stuff is great too.

**Comment by Thorin-Oakenpants** [comment]
- there may only be one sandboxing pref (for now), but it needs to be in the js. Just like we have other items to dispel bad advice from over the years, some of which permeate the web and refuse to die, such as disabling IPv6.
- e10s and sandboxing (1 item) ... renamed it without the sandboxing, but it belongs in this section IMO
- the two new prefs in FF53, I was building them in, and fileURI was already there. I see no point in removing one to keep an eye on but keeping the other
- numbering: renumbered 51-55 (stuck shim as last item since it will be the first and maybe only one to be deprecated), made sandboxing 60 to leave a little room and separate it as a distinct item
- pref( -> user_pref( thanks

So where do we go from here? Draft edited above. I am open to moving the section elsewhere, but where? 0001 is "start" in PB mode, 0100: is "startup" .. and e10s (and I would guess the rest) require a restart - seemed logical to me. I don't think this is personal at all - this is about security.

**Comment by earthlng** [comment]
> Just like we have other items to dispel bad advice

Yeah, fair enough, but we now have a github page for the user.js and IMO that item doesn't need to be in the user.js, but whatever. just my 2 cents

> I see no point in removing one to keep an eye on but keeping the other

One is a fairly "simple" local file isolation that's almost never used and doesn't really matter much if something is not quite ready yet or breaks, while the other is probably a pretty complex addon isolation with tons of potential problems. I'm not even using a single WebExtension addon yet.

> stuck shim as last item since it will be the first and maybe only one to be deprecated

:thumbsup:

> I don't think this is personal at all

It's definitely about security but as I said, thinking ahead and once FF57 hits we end up with a section at the top of the file that nobody really needs to change anymore. Thats my only concern.

**Comment by Thorin-Oakenpants** [comment]
:) Just because we have a github doesn't mean we should split info out info into two places. I want it all in one place and that is local, in the file.

So suggest a numbered section (before shutdown section)... my only concern with it being 0050 was that only one item is active, and when looking at the js, that just creates a whole lotta commented out lines until we hit section 0100 at approx line 80 and start to hit the "meat and potatoes" and people get excited .. 

what about 1050's, 950's, 700's, 2750's ? Pick something .. no hurry, I think we need to wait for 53 to land to confirm 53 won't enforce multiprocessCompatible flag .. and 53 is now delayed until the 26th AFAIK

**Comment by earthlng** [comment]
1100 seems fine. But I don't like making containers 1150 because they're independent from e10s.
Since containers don't seem to work with PB (for now?), containers could be more towards the top of the file, because if someone enables PB-only mode the container prefs are irrelevant. So containers could be 0050.

**Comment by Thorin-Oakenpants** [comment]
i know they're independent, I'm just trying to make things group more logically, and we need two sections, and cache, e10s and containers all lump together nicely as processes i guess. Would hate to waste a whole hundred section on 6 numbered items :) Then again ... we can always renumber - eg shove 2800 shutdown to 4800 and 3000 to 5000 .. maybe move a few sections and voila .. a ton of room.

**Comment by earthlng** [comment]
> 53 is now delayed until the 26th AFAIK

not according to the RapidRelease page. Maybe the 26th is just for a new Dev Edition (now based on Beta). Idk where that date is coming from tbh

> 3000 to 5000

Well, I guess Personal will always be the last section so if we renumber we should probably renumber it to 8000

**Comment by earthlng** [comment]
I think we can go with your draft but I'd add a note about 1103 maybe not being quite ready yet, and for 1110 maybe add a newline after `DO NOT MEDDLE WITH THESE`

**Comment by earthlng** [comment]
I really don't like adding prefs that we recommend to not meddle with. What if we create a new section for such prefs, I'd be okay with that. Somewhere towards the end preferably

**Comment by earthlng** [comment]
Are the syntax errors in 1110 on purpose?

**Comment by Thorin-Oakenpants** [comment]
is that what u meant? the closing */ : https://github.com/ghacksuserjs/ghacks-user.js/commit/9b65d9788d1bb94e72af7180d1821753ee35395c

**Comment by earthlng** [comment]
> is that what u meant?

nope, that's not a syntax error, but [these](https://github.com/ghacksuserjs/ghacks-user.js/pull/95) are
